internal/controllers/machinedeployment/machinedeployment_canupdatemachineset.go

```diff
@@ -0,0 +1,359 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+func matchesMachineSet(req *runtimehooksv1.CanUpdateMachineSetRequest) (bool, []string, error) {
+	var reasons []string
+	match, diff, err := matchesMachineSetSpec(&req.Current.MachineSet, &req.Desired.MachineSet)
+	if err != nil {
+		return false, nil, errors.Wrapf(err, "failed to match MachineSet")
+	}
+	if !match {
+		reasons = append(reasons, fmt.Sprintf("MachineSet cannot be updated in-place: %s", diff))
+	}
+	if req.Current.BootstrapConfigTemplate.Object != nil && req.Desired.BootstrapConfigTemplate.Object != nil {
+		match, diff, err = matchesUnstructuredSpec(req.Current.BootstrapConfigTemplate, req.Desired.BootstrapConfigTemplate)
+		if err != nil {
+			return false, nil, errors.Wrapf(err, "failed to match BootstrapConfigTemplate")
+		}
+		if !match {
+			reasons = append(reasons, fmt.Sprintf("BootstrapConfigTemplate cannot be updated in-place: %s", diff))
+		}
+	}
+	match, diff, err = matchesUnstructuredSpec(req.Current.InfrastructureMachineTemplate, req.Desired.InfrastructureMachineTemplate)
+	if err != nil {
+		return false, nil, errors.Wrapf(err, "failed to match %s", req.Current.InfrastructureMachineTemplate.Object.GetObjectKind().GroupVersionKind().Kind)
+	}
+	if !match {
+		reasons = append(reasons, fmt.Sprintf("%s cannot be updated in-place: %s", req.Current.InfrastructureMachineTemplate.Object.GetObjectKind().GroupVersionKind().Kind, diff))
+	}
+
+	if len(reasons) > 0 {
+		return false, reasons, nil
+	}
+
+	return true, nil, nil
+}
+
+func matchesMachineSetSpec(patched, desired *clusterv1.MachineSet) (equal bool, diff string, matchErr error) {
+	// Note: Wrapping MachineSet specs in a MachineSet for proper formatting of the diff.
+	cleanedUpPatchedMachineSet := &clusterv1.MachineSet{
+		Spec: clusterv1.MachineSetSpec{
+			Template: clusterv1.MachineTemplateSpec{
+				Spec: patched.Spec.Template.Spec,
+			},
+		},
+	}
+	cleanedUpDesiredMachineSet := &clusterv1.MachineSet{
+		Spec: clusterv1.MachineSetSpec{
+			Template: clusterv1.MachineTemplateSpec{
+				Spec: desired.Spec.Template.Spec,
+			},
+		},
+	}
+
+	// Cleanup fields that are not the responsibility of the in-place update extension.
+	// Remove in-place mutable fields.
+	cleanedUpPatchedMachineSet.Spec.Template = *mdutil.MachineTemplateDeepCopyRolloutFields(&cleanedUpPatchedMachineSet.Spec.Template)
+	cleanedUpDesiredMachineSet.Spec.Template = *mdutil.MachineTemplateDeepCopyRolloutFields(&cleanedUpDesiredMachineSet.Spec.Template)
+	// Set refs equal.
+	cleanedUpPatchedMachineSet.Spec.Template.Spec.Bootstrap.ConfigRef = cleanedUpDesiredMachineSet.Spec.Template.Spec.Bootstrap.ConfigRef
+	cleanedUpPatchedMachineSet.Spec.Template.Spec.InfrastructureRef = cleanedUpDesiredMachineSet.Spec.Template.Spec.InfrastructureRef
+	return compare.Diff(cleanedUpPatchedMachineSet, cleanedUpDesiredMachineSet)
```

Comment:
@fabriziopandini Let's talk about this on Monday

I think we should filter out fields here that are not relevant for the diff and we should share a util across here, KCP and the regular MD rollout decision as far as possible.

An alternative would be to explicitly only compare Version / FailureDomain but I think that has the risk that we forget  to add new fields here in the future. In that case I would prefer to get a full rollout instead of an in-place rollout that doesn't work

Note to myself:
* Also update KCP + add test coverage for this logic in both places

Reply:
Pushed a commit to address this